src/hotspot/share/oops/klass.hpp

```diff
@@ -511,18 +511,19 @@ class Klass : public Metadata {
     return (BasicType) btvalue;
   }
 
-  // Want a pattern to quickly diff against layout header in register
-  // find something less clever!
+  // Find the right-most non-zero (e.g., ...1000) bit of the diff of array-of-boolean and array-of-byte layout helpers.
   static int layout_helper_boolean_diffbit() {
-    jint zlh = array_layout_helper(T_BOOLEAN);
-    jint blh = array_layout_helper(T_BYTE);
-    assert(zlh != blh, "array layout helpers must differ");
-    int diffbit = 1;
-    while ((diffbit & (zlh ^ blh)) == 0 && (diffbit & zlh) == 0) {
-      diffbit <<= 1;
-      assert(diffbit != 0, "make sure T_BOOLEAN has a different bit than T_BYTE");
-    }
-    return diffbit;
+    uint zlh = checked_cast<uint>(array_layout_helper(T_BOOLEAN));
+    uint blh = checked_cast<uint>(array_layout_helper(T_BYTE));
+    // get all the bits that are set in zlh and clear in blh
+    uint candidates = (zlh & ~blh);
+    assert(candidates != 0, "must be"); // must be some if there is a solution.
+    // Use well known bit hack to isolate the low bit of candidates.
+    // The usual form is (x & -x), but VS warns (C4146) about unary minus of unsigned.
+    // So use alternate form of negation to avoid warning.
+    uint result = candidates & (~candidates + 1);
+    assert(((result - 1) & result) == 0, "post-condition");
```

Comment:
Maybe use "must be power of 2" instead of  "post-condition".  Also, this value is never going to change.  Can we make the function `constexpr`?

Reply:
Assert's message updated.
`constexpr` cannot be used unless all the called functions (alh and its descendents) are also `constexpr`.